src/Microsoft.Diagnostics.Monitoring.WebApi/Utilities/TraceUtilities.cs

```diff
@@ -97,5 +99,49 @@ public static async Task CaptureTraceAsync(TaskCompletionSource<object> startCom
 
             await pipeProcessor.RunAsync(token);
         }
+
+        public static async Task CaptureTraceUntilEventAsync(TaskCompletionSource<object> startCompletionSource, IEndpointInfo endpointInfo, MonitoringSourceConfiguration configuration, TimeSpan timeout, Stream outputStream, string providerName, string eventName, IDictionary<string, string> payloadFilter, ILogger logger, CancellationToken token)
+        {
+            DiagnosticsClient client = new(endpointInfo.Endpoint);
+            TaskCompletionSource<object> stoppingEventHitSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+            using IDisposable registration = token.Register(
+                () => stoppingEventHitSource.TrySetCanceled(token));
+
+            await using EventTracePipeline pipeProcessor = new(client, new EventTracePipelineSettings
+            {
+                Configuration = configuration,
+                Duration = timeout,
+            },
+            async (eventStream, token) =>
+            {
+                startCompletionSource?.TrySetResult(null);
+                await using EventMonitoringPassthroughStream eventMonitoringStream = new(
+                    providerName,
+                    eventName,
+                    payloadFilter,
+                    onEvent: (traceEvent) =>
+                    {
+                        logger.StoppingTraceEventHit(traceEvent);
+                        stoppingEventHitSource.TrySetResult(null);
+                    },
+                    onPayloadFilterMismatch: logger.StoppingTraceEventPayloadFilterMismatch,
+                    eventStream,
+                    outputStream,
+                    DefaultBufferSize,
+                    callOnEventOnlyOnce: true,
+                    leaveDestinationStreamOpen: true /* We do not have ownership of the outputStream */);
+
+                await eventMonitoringStream.ProcessAsync(token);
+            });
+
+            Task pipelineRunTask = pipeProcessor.RunAsync(token);
+            await Task.WhenAny(pipelineRunTask, stoppingEventHitSource.Task).Unwrap();
+
+            if (stoppingEventHitSource.Task.IsCompleted)
+            {
+                await pipeProcessor.StopAsync(token);
```

Comment:
There's possibly a slight race condition here where the stopping event is noticed, the trace starts rundown due to the duration backstop, and calling trying to stop the trace here if the duration expires between when `stoppingEventHitSource` and this `StopAsync` are executed. In actual use, this is fine. However, the tests that are expecting no rundown to be produced when stopping due to the stopping event may fail randomly if this condition is encountered. I'm not sure if there is anything we can do from a functional test standpoint.